sdk/nodejs/resource.ts

```diff
@@ -179,14 +179,14 @@ export abstract class Resource {
      * The set of providers to use for child resources. Keyed by package name (e.g. "aws").
      */
     // tslint:disable-next-line:variable-name
-    private readonly __providers: Record<string, ProviderResource>;
+    private readonly __providers: Record<string, ProviderResourceOrRef>;
 
     public static isInstance(obj: any): obj is Resource {
         return utils.isInstance<Resource>(obj, "__pulumiResource");
     }
 
     // getProvider fetches the provider for the given module member, if any.
-    public getProvider(moduleMember: string): ProviderResource | undefined {
+    public getProvider(moduleMember: string): ProviderResourceOrRef | undefined {
```

Comment:
@lukehoban this can technically impact people. It's a source break, but not a binary break.  Specifically, you might imagine someone writing:

```ts
const v: ProviderResource | undefined = res.getProvider(...);
```

Because this returns a broader set of values, this can cause a source break later.  Note: it would not be a binary break because this will still only ever return `ProviderResource` instances until the user *explicitly* starts passing along `ProviderRef`s.

Reply:
Ah, that's very nice.  I will do that!

Reply:
So going and trying to do this, I realized we were effectively wanting ProviderRef to effectively appear to be a Provider to old code.  So i took a new approach here whereby we don't have ProviderRef at all, and we instead merge its data into Provider itself.  This greatly decreases the side of hte sdk change.   

For users that are still running into deasync issues the solution doesn't have to be: use this whole new parallel construct that you get with `await ProviderResource.getRef(provider)`.  Instead, it is just do `await ProviderResource.register(provider)`.  So effectively the same net impact on them, but without an impact on the SDK.